File: codechef/ltime30_treeprod.py
import sys
sys.setrecursionlimit(10000)
range = xrange; input = raw_input


def miller_rabin(n):
    """ primality Test
        if n < 3,825,123,056,546,413,051, it is enough to test
        a = 2, 3, 5, 7, 11, 13, 17, 19, and 23.
        Complexity: O(log^3 n)
    """
    assert(n >= 1)
    if n == 2:
        return True
    if n <= 1 or not n & 1:
        return False

    primes = [2, 3, 5, 7, 11, 13, 17, 19, 23]

    d = n - 1
    s = 0
    while not d & 1:
        d >>= 1
        s += 1

    for prime in primes:
        if prime >= n:
            continue
        x = pow(prime, d, n)
        if x == 1:
            continue
        for r in range(s):
            if x == n - 1:
                break
            if r + 1 == s:
                return False
            x = x * x % n
    return True


def dfs(v, p, r, prd, mod):
    ret = 0
    if v > r and prd == 0:
        ret += 1
    for cv, cc in edge[v]:
        if cv != p:
            ret += dfs(cv, v, r, (prd * cc) % mod, mod)
    return ret


def dfs2(v, p, col):
    ret = 1
    color[v] = col
    for c in edge[v]:
        if c != p and not color[c]:
            ret += dfs2(c, v, col)
    return ret
    
N, M = map(int, input().split())
edge = [[] for _ in range(N)]
if N <= 1000:
    for i in range(N-1):
        a, b, c = map(int, input().split())
        a -= 1
        b -= 1
        edge[a].append((b, c % M))
        edge[b].append((a, c % M))

    ans = 0
    for i in range(N):
        ans += dfs(i, -1, i, 1, M)
    print(ans)
elif miller_rabin(M):
    for i in range(N-1):
        a, b, c = map(int, input().split())
        a -= 1
        b -= 1
        if c % M:
            edge[a].append(b)
            edge[b].append(a)
    ans = N * (N - 1) // 2
    color = [0] * N
    cnt = 0
    for i in range(N):
        if not color[i]:
            cnt += 1
            c = dfs2(i, -1, cnt)
            ans -= c * (c - 1) // 2
    print(ans)
else:
    raise Exception

File: codechef/ltime30_treeprod.cpp
#include <bits/stdc++.h>

using namespace std;
typedef long long int ll;
typedef pair<int, int> P;
typedef pair<ll, ll> Pll;
typedef vector<int> Vi;
typedef tuple<int, int, int> T;
#define FOR(i,s,x) for(int i=s;i<(int)(x);i++)
#define REP(i,x) FOR(i,0,x)
#define ALL(c) c.begin(), c.end()
#define DUMP( x ) cerr << #x << " = " << ( x ) << endl

const int dr[4] = {-1, 0, 1, 0};
const int dc[4] = {0, 1, 0, -1};

int N, M;
vector<P> edge[100010];
int color[100010];

bool is_prime(int n) {
  if (n < 2) return false;
  for (int i = 2; i * i <= n; i++) {
    if (n % i == 0) return false;
  }
  return true;
}

int dfs(int v, int p, int r, int prd) {
  int ret = 0;
  assert(M != 0);
  if (v > r && prd == 0) ret++;
  for (P c : edge[v]) {
    int cv = c.first, cc = c.second;
    if (p != cv) ret += dfs(cv, v, r, (prd * cc) % M);
  }
  return ret;
}

ll dfs2(int v, int p, int col) {
  ll ret = 1;
  color[v] = col;
  for (P c : edge[v]) {
    int cv = c.first;
    if (cv != p && !color[cv]) ret += dfs2(cv, v, col);
  }
  return ret;
}

int main() {
  // use scanf in CodeForces!
  cin.tie(0);
  ios_base::sync_with_stdio(false);

  //REP(i, 100) if (is_prime(i)) cout << i << ' ';
  //cout << endl;
  cin >> N >> M;
  if (is_prime(M)) {
    REP(i, N-1) {
      int a, b, c;
      cin >> a >> b >> c; a--, b--;
      if (c % M != 0) {
	edge[a].push_back(P(b, c));
	edge[b].push_back(P(a, c));
      }
    }
    ll n = N, ans = n * (n - 1) / 2;
    int cnt = 0;
    REP(i, N) {
      if (!color[i]) {
	ll c = dfs2(i, -1, ++cnt);
	assert(c >= 0 && c * (c - 1) / 2LL >= 0);
	ans -= c * (c - 1) / 2LL;
      }
    }
    cout << ans << endl;
  } else if (N <= 1000) {
    REP(i, N - 1) {
      int a, b, c;
      cin >> a >> b >> c; a--, b--;
      assert(a < N && b < N);
      edge[a].push_back(P(b, c % M));
      edge[b].push_back(P(a, c % M));
    }

    int ans = 0;
    REP(i, N) ans += dfs(i, -1, i, 1);
    cout << ans << endl;
  }

  return 0;
}
